add "allowHexadecimal" reader option

Caldfir opened the pull request:
Adds optional read-support for hexadecimal numbers matching `0x[0-9a-fA-F]{1,16}` (size depends on LargestUInt). New capability is tied to the **allowHexadecimal** reader-option.

Not sure if there is much interest aside from myself. This would contribute to possible future [JSON5](https://spec.json5.org/) support which does seem to have been considered (https://github.com/open-source-parsers/jsoncpp/issues/209 https://github.com/open-source-parsers/jsoncpp/issues/723). Would need to support leading `(+|-)` for this to be compliant with the JSON5 spec.

BillyDonahue commented on src/lib_json/json_reader.cpp:
```diff
@@ -1639,6 +1671,41 @@ bool OurReader::decodeNumber(Token& token, Value& decoded) {
+    else if (c >= '0')
+      c -= '0';
+    else
+      return addError("Contains non-hexadecimal digits.", token, current);
```
seems that this wouldn't catch nonhex characters like 'z'.

Caldfir replied:
Hmm... I think I had assumed that token.start/token.end were being set correctly by `readHexadecimal()`. It would make the method more robust and flexible to check both bounds though. Will fix.

baylesj commented:
Closing this due to inactivity. Feel free to reopen if you are able to address feedback.

Caldfir commented:
@baylesj The commit from Feb 2024 ([7a74a99](https://github.com/open-source-parsers/jsoncpp/pull/1510/commits/7a74a996f3892f04bb33c052e28b9237fdabf45a)) addressed the comments from @BillyDonahue 

Should the issue be reopened?
